newrelic/hooks/mlmodel_sklearn.py

```diff
@@ -12,12 +12,81 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import sys
+
 from newrelic.api.function_trace import FunctionTrace
 from newrelic.api.time_trace import current_trace
 from newrelic.api.transaction import current_transaction
 from newrelic.common.object_wrapper import wrap_function_wrapper
 
 METHODS_TO_WRAP = ("predict", "fit", "fit_predict", "predict_log_proba", "predict_proba", "transform", "score")
+METRIC_SCORERS = (
+    "accuracy_score",
+    "balanced_accuracy_score",
+    "f1_score",
+    "precision_score",
+    "recall_score",
+    "roc_auc_score",
+    "r2_score",
+)
+PY2 = sys.version_info[0] == 2
+
+
+def _wrap_predict_return_type(data, model_name):
+    """
+    Returns a new NR custom type with model info attached to it as attrs.
+    """
+    import numpy as np
+
+    try:
+        # Numpy NDArrays require special implementation to subclass.
+        if isinstance(data, np.ndarray):
+
+            class NRNumpyNDArray(np.ndarray):
+                _nr_wrapped_model_name = model_name
+
+                def __array_wrap__(self, obj):
+                    if obj.shape == ():
+                        return obj[()]  # if ufunc output is scalar, return it
+                    else:
+                        return np.ndarray.__array_wrap__(self, obj)
+
+            return data.view(NRNumpyNDArray)
+
+        # Booleans are singletons and require special implementation to subclass.
+        if isinstance(data, (bool)):
+            if PY2:
+
+                class NRBoolType:
+                    _nr_wrapped_model_name = model_name
+
+                    def __init__(self, value):
+                        self.value = value
+
+                    def __nonzero__(self):
+                        return bool(self.value)
+
+            else:
+
+                class NRBoolType:
+                    _nr_wrapped_model_name = model_name
+
+                    def __init__(self, value):
+                        self.value = value
+
+                    def __bool__(self):
+                        return bool(self.value)
+
+            return NRBoolType(data)
+
+        # Attempt to subclass the type directly.
+        class NRWrapType(type(data)):
+            _nr_wrapped_model_name = model_name
+
+        return NRWrapType(data)
+    except:  # Squash any exceptions resulting from attempted wrap typing.
+        pass
+    return data
```

Comment:
This function is super brittle and dangerous, I'm not comfortable with the implementation here. If all we need is an object that we can set attributes on that can be achieved much more easily with an ObjectProxy. No need to make custom subclasses with custom compatibility functionality required.

I'll push up a commit to swap this out for a call to ObjectProxy and you can take a look at it and see if you agree.